File: examples_rust/examples/example_random.rs
use libzupt_rust;

fn main() {
    println!("{}", "=".repeat(60));
    println!("libzupt - Random Bytes and Hashing Example (Rust)");
    println!("{}", "=".repeat(60));
    println!();

    println!("1. Generating random bytes...");
    let random_bytes = libzupt_rust::random_bytes(32);
    println!("   Generated {} random bytes:", random_bytes.len());
    println!("   {}", hex::encode(&random_bytes));
    println!();

    println!("2. Generating AES nonce...");
    let nonce = libzupt_rust::random_bytes(libzupt_rust::AES_NONCE_SIZE);
    println!("   Nonce ({} bytes): {}", nonce.len(), hex::encode(&nonce));
    println!();

    let data = b"Hello, Post-Quantum World!";

    println!("3. Computing SHA-256 hash...");
    let sha256_hash = libzupt_rust::sha256(data);
    println!("   Data: {}", String::from_utf8_lossy(data));
    println!("   SHA-256: {}", hex::encode(&sha256_hash));
    println!();

    println!("4. Computing SHA3-512 hash...");
    let sha3_512_hash = libzupt_rust::sha3_512(data);
    println!("   Data: {}", String::from_utf8_lossy(data));
    println!("   SHA3-512: {}", hex::encode(&sha3_512_hash));
    println!();

    println!("5. Simulating key derivation...");
    let salt = libzupt_rust::random_bytes(16);
    println!("   Salt: {}", hex::encode(&salt));
    let mut input = salt.clone();
    input.extend_from_slice(b"my-secret-password");
    let derived_key = libzupt_rust::sha256(&input);
    println!("   Derived key (32 bytes): {}", hex::encode(&derived_key));
    println!();

    println!("{}", "=".repeat(60));
    println!("Random bytes and hashing example passed!");
    println!("{}", "=".repeat(60));
}

File: examples_go/example_random.go
package main

import (
	"encoding/hex"
	"fmt"
	"strings"
)

func runExampleRandom() error {
	sep := strings.Repeat("=", 60)
	fmt.Println(sep)
	fmt.Println("libzupt - Random Bytes and Hashing Example")
	fmt.Println(sep)
	fmt.Println()

	fmt.Println("1. Generating random bytes...")
	randomBytes := RandomBytes(32)
	fmt.Printf("   Generated %d random bytes:\n", len(randomBytes))
	fmt.Printf("   %s\n", hex.EncodeToString(randomBytes))
	fmt.Println()

	fmt.Println("2. Generating AES nonce...")
	nonce := RandomBytes(AESNonceSize)
	fmt.Printf("   Nonce (%d bytes): %s\n", len(nonce), hex.EncodeToString(nonce))
	fmt.Println()

	fmt.Println("3. Computing SHA-256 hash...")
	data := []byte("Hello, Post-Quantum World!")
	sha256Hash := Sha256(data)
	fmt.Printf("   Data: %s\n", data)
	fmt.Printf("   SHA-256: %s\n", hex.EncodeToString(sha256Hash))
	fmt.Println()

	fmt.Println("4. Computing SHA3-512 hash...")
	sha3_512Hash := Sha3_512(data)
	fmt.Printf("   Data: %s\n", data)
	fmt.Printf("   SHA3-512: %s\n", hex.EncodeToString(sha3_512Hash))
	fmt.Println()

	fmt.Println("5. Simulating key derivation...")
	salt := RandomBytes(16)
	fmt.Printf("   Salt: %s\n", hex.EncodeToString(salt))
	derivedInput := make([]byte, 0, len(salt)+len("my-secret-password"))
	derivedInput = append(derivedInput, salt...)
	derivedInput = append(derivedInput, []byte("my-secret-password")...)
	derivedKey := Sha256(derivedInput)
	fmt.Printf("   Derived key (32 bytes): %s\n", hex.EncodeToString(derivedKey))
	fmt.Println()

	fmt.Println(sep)
	fmt.Println("Random bytes and hashing example passed!")
	fmt.Println(sep)
	return nil
}
